### ISSUE TITLE
Performance issue with multiple ingressgateway

### ISSUE THREAD

**Issue by pandey-adarsh147** [istio/istio#23746]
**Bug description**
If one instance of ingressgateway serving 'x' number of request, then two instance should server '2x' number of request. Currently, it is serving 'x/2'. Number of request getting decrease to constant, doesn't matter how many number of instance I will increase. (gRPC request)

**Expected behavior**
If one instance of ingressgateway serving 'x' number of request, then two instance should server '2x' number of request. Currently, it is serving 'x/2'. Number of request getting decrease to constant, doesn't matter how many number of instance I will increase. (gRPC request)

**Steps to reproduce the bug**
Create ingressgateway on 3 node cluster, deploy service and call it from outside cluster. I am using gzh tool to put load.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.7", GitCommit:"65ecaf0671341311ce6aea0edab46ee69f65d59e", GitTreeState:"clean", BuildDate:"2019-01-24T19:32:00Z", GoVersion:"go1.10.7", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.10", GitCommit:"059c666b8d0cce7219d2958e6ecc3198072de9bc", GitTreeState:"clean", BuildDate:"2020-04-03T15:17:29Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```
```
Istio version 1.5.2 and 1.1.3

istioctl version
Version: 1.0.1
GitRevision: 42773aacced474d97159902d20579a25b1f98106
User: root@832d5020b1d4
Hub: gcr.io/istio-release
GolangVersion: go1.10.1
BuildStatus: Clean
```

**How was Istio installed?**
Helm chart

**Environment where bug was observed (cloud vendor, OS, etc)**
Azure AKS, Kubernetes v1.15.10 and v1.14.8

Load test report
```
With single ingressgateway instance (ilbgateway)
Summary:
  Count:	98713
  Total:	60.01 s
  Slowest:	9.88 s
  Fastest:	31.24 ms
  Average:	182.19 ms
  Requests/sec:	1645.00

Response time histogram:
  31.235 [1]	|
  1016.102 [96750]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  2000.968 [1194]	|
  2985.834 [135]	|
  3970.701 [74]	|
  4955.567 [66]	|
  5940.433 [74]	|
  6925.300 [114]	|
  7910.166 [3]	|
  8895.033 [0]	|
  9879.899 [1]	|

Latency distribution:
  10 % in 51.73 ms
  25 % in 56.76 ms
  50 % in 69.59 ms
  75 % in 168.93 ms
  90 % in 497.63 ms
  95 % in 597.63 ms
  99 % in 1.32 s

Status code distribution:
  [OK]                 98412 responses
  [DeadlineExceeded]   1 responses
  [Unavailable]        300 responses
```

------------------------------------------------------

```
With 3 instance of ingressgateway (ilbgateway)
Summary:
  Count:	46128
  Total:	60.01 s
  Slowest:	3.06 s
  Fastest:	31.47 ms
  Average:	389.99 ms
  Requests/sec:	768.70

Response time histogram:
  31.466 [1]	|
  334.649 [30673]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  637.832 [4136]	|∎∎∎∎∎
  941.015 [4066]	|∎∎∎∎∎
  1244.198 [3362]	|∎∎∎∎
  1547.381 [2108]	|∎∎∎
  1850.564 [907]	|∎
  2153.747 [422]	|∎
  2456.930 [144]	|
  2760.113 [5]	|
  3063.296 [4]	|

Latency distribution:
  10 % in 51.07 ms
  25 % in 71.91 ms
  50 % in 140.37 ms
  75 % in 606.90 ms
  90 % in 1.13 s
  95 % in 1.39 s
  99 % in 1.94 s

Status code distribution:
  [OK]            45828 responses
  [Unavailable]   300 responses
```

**Comment by howardjohn** [comment]
> istioctl version
> Version: 1.0.1
> GitRevision: 42773aacced474d97159902d20579a25b1f98106
> User: root@832d5020b1d4
> Hub: gcr.io/istio-release
> GolangVersion: go1.10.1
> BuildStatus: Clean

Are you using Istio 1.0.1? Or is this just the version of istioctl? Either way -- please upgrade to a more recent version

**Comment by pandey-adarsh147** [comment]
I am using istio version 1.5.2. Also, confirmed v1.1.3

**Comment by howardjohn** [comment]
So you see on 1.5 and 1.1?

**Comment by howardjohn** [comment]
Is your test app public so we can reproduce?

**Comment by pandey-adarsh147** [comment]
Yes v1.5.2 and v1.1. 
No it is not public, but it is reproducible on any istio cluster.

**Comment by howardjohn** [comment]
I have done a fair amount of performance testing and have not sent his. It would be really helpful if you had a reproducable example as there are many variables that could impact this

**Comment by thejasn** [comment]
Will work on getting a reproducible example. But another question related to this anomaly, the above stats was observed while using the istio-ilbgateway, but performing the same tests on istio-ingressgateway results in expected behavior.

With single instance of istio-ingressgateway,
```
Summary:
  Count:	87320
  Total:	60.00 s
  Slowest:	1.25 s
  Fastest:	40.43 ms
  Average:	68.66 ms
  Requests/sec:	1455.32Response time histogram:
  40.435 [1]	|
  161.046 [85840]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  281.657 [301]	|
  402.269 [469]	|
  522.880 [307]	|
  643.491 [99]	|
  764.103 [138]	|
  884.714 [42]	|
  1005.326 [13]	|
  1125.937 [1]	|
  1246.548 [10]	|Latency distribution:
  10 % in 49.73 ms 
  25 % in 54.46 ms 
  50 % in 61.72 ms 
  75 % in 70.08 ms 
  90 % in 78.81 ms 
  95 % in 86.54 ms 
  99 % in 334.70 ms Status code distribution:
  [OK]            87221 responses   
  [Unavailable]   99 responses      Error distribution:
  [99]   rpc error: code = Unavailable desc = transport is closing
``` 


With 3 instances of ingressgateway,
```
Summary:
  Count:	92485
  Total:	60.00 s
  Slowest:	1.29 s
  Fastest:	39.86 ms
  Average:	64.82 ms
  Requests/sec:	1541.40Response time histogram:
  39.861 [1]	|
  164.515 [91893]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  289.169 [187]	|
  413.823 [162]	|
  538.478 [124]	|
  663.132 [3]	|
  787.786 [15]	|
  912.440 [0]	|
  1037.094 [0]	|
  1161.748 [0]	|
  1286.402 [1]	|Latency distribution:
  10 % in 50.58 ms 
  25 % in 55.16 ms 
  50 % in 61.97 ms 
  75 % in 69.84 ms 
  90 % in 78.21 ms 
  95 % in 84.09 ms 
  99 % in 105.33 ms Status code distribution:
  [OK]            92386 responses   
  [Unavailable]   99 responses      Error distribution:
  [99]   rpc error: code = Unavailable desc = transport is closing
```
So are the two gateways that different in their implementations ??

**Comment by howardjohn** [comment]
By ilb you mean you out `service.beta.kubernetes.io/azure-load-balancer-internal` annotation right?

If that's the difference between fast and slow it may be an AKS issue potentially

**Comment by pandey-adarsh147** [comment]
During installation. We are enabling gateways.istio-ilbgateway.enabled.

**Comment by anand5555** [comment]
I am also using aks internal loadbalancer and  seeing higher latency for ingress gateway upto 9-10 sec . I have tried it with a public load Balancer and i see the same response times . @howardjohn . Any thing else i can try ?

Istio version : 1.2.5
aks version : 1.14.7

**Comment by howardjohn** [comment]
If its an issue with ILB *only* that strongly suggests that the issue is in AKS, not Istio. Note the LB/ILB happens before the packet even reaches istio-ingressgateway pod, so there is nothing we can do.

One way to prove its the ILB:

* Do `load-client http://ILB-IP` and observe it is slow as scale increases
* Do `load-client http://istio-ingressgateway.istio-system` and observe if it slows down as scale increases

Basically this bypasses the ILB/LB. I strongly suspect you will not see a performance impact if you do this.

**Comment by howardjohn** [comment]
Any more info on ^?

**Comment by anand5555** [comment]
In our case ... Lately we found that, these latencies are caused by AKS because of the linux kernel race condition which are due to DNS lookups .

**Comment by howardjohn** [comment]
I am fairly sure this is an AKS issue, not Istio, and there has been no followup so I will close this. If we find evidence suggesting otherwise let me know.
